components/webgpu/canvas_context.rs

```diff
@@ -0,0 +1,771 @@
+        };
+        let copy_size = wgt::Extent3d {
+            width: config.size.width,
+            height: config.size.height,
+            depth_or_array_layers: 1,
+        };
+        self.global.command_encoder_copy_texture_to_buffer(
+            encoder_id,
+            &texture_info,
+            &buffer_info,
+            &copy_size,
+        )?;
+        let (command_buffer_id, error) = self
+            .global
+            .command_encoder_finish(encoder_id, &wgt::CommandBufferDescriptor::default());
+        if let Some(error) = error {
+            return Err(error.into());
+        };
+        Ok(command_buffer_id)
+    }
+
+    /// Unmaps buffer or cancel mapping if in progress
+    fn unmap(&mut self) {
+        match self.state {
+            StagingBufferState::Unassigned | StagingBufferState::Available(_) => {},
+            StagingBufferState::Mapping(buffer) |
+            StagingBufferState::Mapped(MappedBuffer { buffer, .. }) => {
+                let _ = self.global.buffer_unmap(self.buffer_id);
+                self.state = StagingBufferState::Available(buffer)
+            },
+        }
+    }
+
+    /// Obtains snapshot from mapped buffer
+    fn snapshot(&self) -> Option<Snapshot> {
+        let StagingBufferState::Mapped(mapped) = &self.state else {
+            return None;
+        };
+        let format = match mapped.image_format {
+            ImageFormat::RGBA8 => SnapshotPixelFormat::RGBA,
+            ImageFormat::BGRA8 => SnapshotPixelFormat::BGRA,
+            _ => unimplemented!(),
+        };
+        let alpha_mode = if mapped.is_opaque {
+            SnapshotAlphaMode::AsOpaque {
+                premultiplied: false,
+            }
+        } else {
+            SnapshotAlphaMode::Transparent {
+                premultiplied: true,
+            }
+        };
+        let padded_byte_width = mapped.stride();
+        let data = mapped.slice();
+        let bytes_per_pixel = mapped.image_format.bytes_per_pixel() as usize;
+        let mut result_unpadded =
+            Vec::<u8>::with_capacity(mapped.image_size.area() as usize * bytes_per_pixel);
+        for row in 0..mapped.image_size.height {
+            let start = (row * padded_byte_width).try_into().ok()?;
+            result_unpadded
+                .extend(&data[start..start + mapped.image_size.width as usize * bytes_per_pixel]);
+        }
+        let mut snapshot =
+            Snapshot::from_vec(mapped.image_size, format, alpha_mode, result_unpadded);
+        if mapped.is_opaque {
+            snapshot.transform(SnapshotAlphaMode::Opaque, snapshot.format())
+        }
+        Some(snapshot)
+    }
+}
+
+impl Drop for StagingBuffer {
+    fn drop(&mut self) {
+        match self.state {
+            StagingBufferState::Unassigned => {},
+            StagingBufferState::Available(_) |
+            StagingBufferState::Mapping(_) |
+            StagingBufferState::Mapped(_) => {
+                self.global.buffer_drop(self.buffer_id);
```

Comment:
I assume this cannot happen during a mapping operation? Might be worth putting an `unreachable` if that's the case.